### PR TITLE
[init refactor] introducing init analyzers

### PR DESCRIPTION
**Pull request by balopat**
Refactoring - shouldn't change anything functionally. 
Introducing analyzers in the `initializer` package. 

An Analyzer is basically implementing a Visitor pattern, can react to EnterDir and ExitDir events and can be called on each file. This creates a more composable design that is open for extension but closed for modification. 

Also added tests to cover config generation logic.

### CODE REVIEW

File: pkg/skaffold/initializer/analyze.go [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+	"github.com/karrick/godirwalk"
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/initializer/kubectl"
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/util"
+)
+
+type analysis struct {
+	kubectlAnalyzer  *KubectlAnalyzer
```
Comment (dgageot) [inline pkg/skaffold/initializer/analyze.go]: Do these three types need to be exposed?

File: pkg/skaffold/initializer/analyze.go [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+func (a *DirectoryAnalyzer) AnalyzeFile(filePath string) error {
+	return nil
+}
+
+func (a *DirectoryAnalyzer) EnterDir(dir string) {
+	a.currentDir = dir
+}
+func (a *DirectoryAnalyzer) ExitDir(dir string) {
```
Comment (dgageot) [inline pkg/skaffold/initializer/analyze.go]: nit: blank line

File: pkg/skaffold/initializer/analyze.go [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+type KubectlAnalyzer struct {
+	DirectoryAnalyzer
+	kubernetesManifests []string
+}
+
+func (a *KubectlAnalyzer) AnalyzeFile(filePath string) error {
+	isSkaffoldConfig := IsSkaffoldConfig(filePath)
+	isKubernetesManifest := kubectl.IsKubernetesManifest(filePath)
+
+	if isKubernetesManifest && !isSkaffoldConfig {
```
Comment (dgageot) [inline pkg/skaffold/initializer/analyze.go]: ni: by inlining those variables, we could avoid computing some computation.

File: pkg/skaffold/initializer/analyze.go [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+type SkaffoldConfigAnalyzer struct {
+	DirectoryAnalyzer
+	force bool
+}
+
+func (a *SkaffoldConfigAnalyzer) AnalyzeFile(filePath string) error {
+	isSkaffoldConfig := IsSkaffoldConfig(filePath)
+	if isSkaffoldConfig {
```
Comment (dgageot) [inline pkg/skaffold/initializer/analyze.go]: nit: what about inverting the test? guard vs nesting

File: pkg/skaffold/initializer/builders.go [hunk 5]
```diff
@@ -218,23 +217,7 @@ func resolveBuilderImages(builderConfigs []InitBuilder, images []string, force b
 		images = util.RemoveFromSlice(images, image)
 	}
 	if len(choices) > 0 {
-		logrus.Warnf("unused builder configs found in repository: %v", choices)
+		logrus.Warnf("unused builder kubernetesManifests found in repository: %v", choices)
```
Comment (dgageot) [inline pkg/skaffold/initializer/builders.go]: Not sure I understand this message, as a user.

File: pkg/skaffold/initializer/config_test.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+*/
+
+package initializer
+
+import (
+	"errors"
+	"path/filepath"
+	"testing"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/build/buildpacks"
```
Comment (dgageot) [inline pkg/skaffold/initializer/config_test.go]: gofmt

File: pkg/skaffold/initializer/analyze.go (hunk 1 above)
Comment (balopat) [reply pkg/skaffold/initializer/analyze.go]: Actually, they don't!

**Comment by balopat** [comment]
Thanks for the quick review! I'll fix the nits tomorrow morning!:)

File: pkg/skaffold/initializer/builders.go (hunk 5 above)
Comment (balopat) [reply pkg/skaffold/initializer/builders.go]: accidental comment refactoring, reverted!

File: pkg/skaffold/initializer/analyze.go (hunk 4 above)
Comment (balopat) [reply pkg/skaffold/initializer/analyze.go]: great point!

**Pull request merged**